b2share/modules/records/api.py

```diff
@@ -0,0 +1,128 @@
+# waive the privileges and immunities granted to it by virtue of its status
+# as an Intergovernmental Organization or submit itself to any jurisdiction.
+
+"""B2Share Deposit API."""
+
+from elasticsearch.exceptions import NotFoundError
+from invenio_db import db
+from invenio_pidstore.resolver import Resolver
+from invenio_pidstore.models import PersistentIdentifier
+from invenio_deposit.api import Deposit
+from invenio_records.models import RecordMetadata
+from invenio_records.api import Record
+from invenio_records_files.api import Record, FilesIterator, FileObject
+from invenio_records_files.utils import sorted_files_from_bucket
+from invenio_files_rest.models import Bucket, ObjectVersion, FileInstance
+from b2share.modules.deposit.fetchers import b2share_deposit_uuid_fetcher
+from invenio_indexer.api import RecordIndexer
+from invenio_pidrelations.contrib.versioning import PIDVersioning
+
+class B2ShareFileObject(FileObject):
+    """Wrapper for B2Share files."""
+
+    def dumps(self):
+        """Dump the record metadata."""
+        if self.obj.file.storage_class == 'B':
+            self.b2safe_pid = True
+        else:
+            self.b2safe_pid = False
+        self.data.update({
+            'bucket': str(self.obj.bucket_id),
+            'checksum': self.obj.file.checksum,
+            'key': self.obj.key,  # IMPORTANT it must stay here!
+            'size': self.obj.file.size,
+            'version_id': str(self.obj.version_id),
+            'b2safe_pid': self.b2safe_pid,
+        })
+        return self.data
+
+
+class B2ShareRecord(Deposit, Record):
+    """B2Share record class."""
+    file_cls = B2ShareFileObject
+
+    @classmethod
+    def delete(cls, pid, record):
+        """Delete a published record.
+
+        params:
+        pid: PersistentIdentifier
+        record: invenio_records_files.api.Record instance
+        """
+        # Fetch deposit id from record and resolve deposit record and pid.
+        depid = b2share_deposit_uuid_fetcher(None, record)
+        if not depid:
+            abort(404)
+        depid, deposit = Resolver(
+            pid_type=depid.pid_type,
+            object_type='rec',
+            # Retrieve the deposit with the Record class on purpose as the current
+            # Deposit api prevents the deletion of published deposits.
+            getter=Record.get_record,
+        ).resolve(depid.pid_value)
+
+        # Note: the record is unindexed from Elasticsearch via the
+        # signal handlers in triggers.py
+        try:
+            RecordIndexer().delete(deposit)
+        except NotFoundError:
+            pass
+
+        # remove record from versioning chain
+        # if the parent has no other children and no draft child
+        # delete it from the db completely
+        version_child = PIDVersioning(child=pid)
+        version_master = PIDVersioning(parent=version_child.parent)
+        if version_master is not None:
+            version_master.remove_child(pid)
```

Comment:
We should keep the link between the parent and the child.
Use `PIDVersioning.update_redirect`.